File: src/scripts/model/citizen.js
import {CommercialZone} from './buildings/zones/commercial.js';
import {IndustrialZone} from './buildings/zones/industrial.js';
import {ResidentialZone} from './buildings/zones/residential.js';
import config from '../config.js';

let nextCitizenId = 0;

export const CitizenState = {
    idle: 'idle',
    school: 'school',
    employed: 'employed',
    unemployed: 'unemployed',
    retired: 'retired',
}

export class Citizen {
    id = nextCitizenId++;

    state = CitizenState.idle;
    stateCounter = 0;
    age = 0;
    /**
     * Reference to the building the citizen lives at
     * @type {ResidentialZone}
     */
    residence = null;

    /**
     * Reference to the building the citizen works at
     * @type {CommercialZone | IndustrialZone}
     */
    workplace = null;

    /**
     * @type {boolean}
     */
    updated = true;

    constructor() {

        this.name = generateRandomName();
        /**
         * Age of the citizen in years
         * @type {number}
         */
        this.age = 1 + Math.floor(100 * Math.random());

        if (this.age < config.citizen.minWorkingAge) {
            this.state = CitizenState.school;
        } else if (this.age >= config.citizen.retirementAge) {
            this.state = CitizenState.retired;
        } else {
            this.state = CitizenState.unemployed;
        }
    }

    /**
     * Returns an HTML representation of this object
     * @returns {string}
     */
    toHTML() {
        return `
      <li class="info-citizen">
        <span class="info-citizen-name">${this.name}</span>
        <br>
        <span class="info-citizen-details">
          <span>
            <img class="info-citizen-icon" src="/icons/calendar.png">
            ${this.age} 
          </span>
          <span>
            <img class="info-citizen-icon" src="/icons/job.png">
            ${this.state}
          </span>
        </span>
      </li>
    `;
    }
}

function generateRandomName() {
    const firstNames = [
        'Emma', 'Olivia', 'Ava', 'Sophia', 'Isabella',
        'Liam', 'Noah', 'William', 'James', 'Benjamin',
        'Elizabeth', 'Margaret', 'Alice', 'Dorothy', 'Eleanor',
        'John', 'Robert', 'William', 'Charles', 'Henry',
        'Alex', 'Taylor', 'Jordan', 'Casey', 'Robin'
    ];

    const lastNames = [
        'Smith', 'Johnson', 'Williams', 'Jones', 'Brown',
        'Davis', 'Miller', 'Wilson', 'Moore', 'Taylor',
        'Anderson', 'Thomas', 'Jackson', 'White', 'Harris',
        'Clark', 'Lewis', 'Walker', 'Hall', 'Young',
        'Lee', 'King', 'Wright', 'Adams', 'Green'
    ];

    const randomFirstName = firstNames[Math.floor(Math.random() * firstNames.length)];
    const randomLastName = lastNames[Math.floor(Math.random() * lastNames.length)];

    return randomFirstName + ' ' + randomLastName;
}

File: Simulation/Model/Citizen.cs
using Simulation.Model.Buildings.Zones;

namespace Simulation.Model;

public static class CitizenState
{
    public const string Idle       = "idle";
    public const string School     = "school";
    public const string Employed   = "employed";
    public const string Unemployed = "unemployed";
    public const string Retired    = "retired";
}

public class Citizen
{
    private static int _nextId;

    public int           Id           { get; } = Interlocked.Increment(ref _nextId);
    public string        Name         { get; }
    public string        State        { get; set; }
    public int           StateCounter { get; set; }
    public int           Age          { get; set; }
    public ResidentialZone? Residence { get; set; }
    public ZoneWithJobs?    Workplace { get; set; }
    public bool          Updated      { get; set; } = true;

    public Citizen()
    {
        Name = NameGenerator.Next();
        Age  = 1 + Random.Shared.Next(100);

        State = Age < SimConfig.Citizen.MinWorkingAge  ? CitizenState.School
              : Age >= SimConfig.Citizen.RetirementAge ? CitizenState.Retired
              :                                          CitizenState.Unemployed;
    }
}

internal static class NameGenerator
{
    private static readonly string[] First =
    [
        "Emma", "Olivia", "Ava", "Sophia", "Isabella",
        "Liam", "Noah", "William", "James", "Benjamin",
        "Elizabeth", "Margaret", "Alice", "Dorothy", "Eleanor",
        "John", "Robert", "Charles", "Henry",
        "Alex", "Taylor", "Jordan", "Casey", "Robin",
    ];

    private static readonly string[] Last =
    [
        "Smith", "Johnson", "Williams", "Jones", "Brown",
        "Davis", "Miller", "Wilson", "Moore", "Taylor",
        "Anderson", "Thomas", "Jackson", "White", "Harris",
        "Clark", "Lewis", "Walker", "Hall", "Young",
        "Lee", "King", "Wright", "Adams", "Green",
    ];

    public static string Next() =>
        $"{First[Random.Shared.Next(First.Length)]} {Last[Random.Shared.Next(Last.Length)]}";
}
